core/state/journal_api.go

```diff
@@ -0,0 +1,86 @@
+package state
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+type journal interface {
+	// snapshot returns an identifier for the current revision of the state.
+	// The lifeycle of journalling is as follows:
+	// - snapshot() starts a 'scope'.
+	// - Tee method snapshot() may be called any number of times.
+	// - For each call to snapshot, there should be a corresponding call to end
+	//  the scope via either of:
+	//   - revertToSnapshot, which undoes the changes in the scope, or
+	//   - discardSnapshot, which discards the ability to revert the changes in the scope.
+	//     - This operation might merge the changes into the parent scope.
+	//       If it does not merge the changes into the parent scope, it must create
+	//       a new snapshot internally, in order to ensure that order of changes
+	//       remains intact.
+	snapshot() int
+
+	// revertToSnapshot reverts all state changes made since the given revision.
+	revertToSnapshot(revid int, s *StateDB)
```

Comment:
revertTOsnapshot and discardSnapshot seems asymmetric (one discards something, the other reverts **up to** something).

Would it make sense to unify these? I.e. have discard also take the parent until which to discard (if revert knows where to revert, discard should also know? shoudln't it be called in the same spot?).

Also, discard doesn't really need an id, does it, you will only ever discard the head?

Reply:
Myeah, I guess you are right. At the moment, though, the `linear_journal` (a.k.a our current journal), doesn't actually do anything on discard: it expects the revert to simply tell it where to go. 

So yeah, for the sake of clarity and semantic cleanliness, I will change that.

Reply:
Right. I have a comment in `journal_set.go`: 
```
	want := len(j.entries) - 1
	have := id
	if want != have {
		if want == 0 && id == 1 {
			// If a transaction is applied successfully, the statedb.Finalize will
			// end by clearing and resetting the journal. Invoking a discardSnapshot
			// afterwards will lead us here.
			// Let's not panic, but it's ok to complain a bit
			log.Error("Extraneous invocation to discard snapshot")
			return
		} else {
			panic(fmt.Sprintf("journalling error, want discard(%d), have discard(%d)", want, have))
		}
	}
```
So, yes, it does expect `discard` only to occur on the last entry. However, it's because the statedb.Finalize does this: 
```
	// Invalidate journal because reverting across transactions is not allowed.
	s.clearJournalAndRefund()
```
So, we apply a transaction. If it succeeds, the statedb will `clear` the journal, making it new again. But, the miner does this: 
```
	var (
		snap = env.state.Snapshot()
		gp   = env.gasPool.Gas()
	)
	receipt, err := core.ApplyTransaction(miner.chainConfig, miner.chain, &env.coinbase, env.gasPool, env.state, env.header, tx, &env.header.GasUsed, vm.Config{})
	if err != nil {
		env.state.RevertToSnapshot(snap)
		env.gasPool.SetGas(gp)
	}
```
So, the miner will potentially call `RevertToSnapshot`. It will _never_ call `Discard`, since that happens internally inside `ApplyTransaction`. From an APi-perspective, this is bad. The `Snapshot()` and `RevertToSnapshot` are collected in one place, but the `Discard` is not.